runtime/compiler/x/codegen/J9TreeEvaluator.cpp

```diff
@@ -10341,6 +10341,473 @@ inlineCompareAndSwapNative(
+   generateRegRegInstruction(TR::InstOpCode::ADD4RegReg, node, loopLimitReg, offsetReg, cg);
+
+
+   generateLabelInstruction(TR::InstOpCode::label, node, loopLabel, cg);
+
+   // if index >= loopLimit, jump to handling the residual bytes
+   generateRegRegInstruction(TR::InstOpCode::CMP4RegReg, node, indexReg, loopLimitReg, cg);
+   generateLabelInstruction(TR::InstOpCode::JGE4, node, residualLabel, cg);
+
+   // Load 16 bytes from address [buf + index]
+   generateRegMemInstruction(TR::InstOpCode::MOVDQURegMem, node, xmmChunkReg, generateX86MemoryReference(bufReg, indexReg, 0, TR::Compiler->om.contiguousArrayHeaderSizeInBytes(), cg), cg);
+
+   // Create zero vector
+   generateRegRegInstruction(TR::InstOpCode::PXORRegReg, node, xmmMaskReg, xmmMaskReg, cg);
+
+   // Compare 0 > chunk bytewise
+   generateRegRegInstruction(TR::InstOpCode::PCMPGTBRegReg, node, xmmMaskReg, xmmChunkReg, cg);
+
+   // Extract bitmask of negative values
+   generateRegRegInstruction(TR::InstOpCode::PMOVMSKB4RegReg, node, maskReg, xmmMaskReg, cg);
+
+   // Check if any negative values exist
+   generateRegRegInstruction(TR::InstOpCode::TEST4RegReg, node, maskReg, maskReg, cg);
+
+   // If the result is nonzero, handle finding a negative byte
+   if (recognizedMethod == TR::java_lang_StringCoding_hasNegatives)
+      {
+      // Jump to returnHasNegativesLabel
+      generateLabelInstruction(TR::InstOpCode::JNE4, node, returnHasNegativesLabel, cg);
+      }
+   else
+      {
+      // Break and calculate index of first negative byte
+      generateLabelInstruction(TR::InstOpCode::JNE4, node, foundNegativeLabel, cg);
+      }
+
+   // increment index by 16 and jump back to the top of the loop
+   generateRegImmInstruction(TR::InstOpCode::ADD4RegImm4, node, indexReg, 16, cg);
+   generateLabelInstruction(TR::InstOpCode::JMP4, node, loopLabel, cg);
+
+
+   // If we found any negative bytes, calculate the index of the first negative byte in the chunk
+   if (recognizedMethod == TR::java_lang_StringCoding_countPositives)
+      {
+      generateLabelInstruction(TR::InstOpCode::label, node, foundNegativeLabel, cg);
+
+      // Calculate the index of the first negative byte within the chunk
+      generateRegRegInstruction(TR::InstOpCode::TZCNT4RegReg, node, maskReg, maskReg, cg);
+
+      // Add the index, which points to the beginning of the chunk,
+      // for the overall index of the first negative
+      generateRegRegInstruction(TR::InstOpCode::ADD4RegReg, node, indexReg, maskReg, cg);
+
+      generateLabelInstruction(TR::InstOpCode::JMP4, node, returnHasNegativesLabel, cg);
+      }
+
+
+   // Deal with the residual (last 15 or fewer) bytes
+   generateLabelInstruction(TR::InstOpCode::label, node, residualLabel, cg);
+
+   // Calculate bytes remaining: loopLimit = -(loopLimit - limit)
+   generateRegRegInstruction(TR::InstOpCode::SUB4RegReg, node, loopLimitReg, limitReg, cg);
+   generateRegInstruction(TR::InstOpCode::NEG4Reg, node, loopLimitReg, cg);
+
+   /*
```

Comment:
Thinking about the handling of the residue bytes for `countPositives`, I wonder whether it could be simplified by doing something like this.

If there are eight or more bytes remaining, load eight bytes, mask the sign bits, and use `lzcnt` to check for negative bytes.  If any were found, update the index, and finish.

For the remaining seven or fewer bytes, load them in chunks, conditionally, based on the number of bytes remaining - four, then two, then one - then use `lzcnt` on all of them.  I can sketch this out in more detail if it's not clear.

Reply:
I think I see what you're saying! By "using `lzcnt` on all of them", do you mean loading all of those bytes into the same register and then using `lzcnt` afterwards? Would that involve shifting after each load to line everything up correctly?

Reply:
I have implemented this processing strategy by first moving to the end of the residual bytes and then working backwards, first possibly loading the last 4 bytes, then 2 (shifting if necessary), then 1 (again, shifting if necessary). This places all of the residual bytes in the correct order in the chunk register, allowing us to calculate the index of any negative bytes.